File: component/php/src/component/phpmetrics.go
/*
Copyright 2016 - 2017 Huawei Technologies Co., Ltd. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package main

import (
	"fmt"
	"os"
	"os/exec"
	"util/git"
	"util/input"
	"util/file"
)

const (
	basePath string = "./workspace"
	// baseCommand string = "phpmetrics"
	baseCommand string = "/home/composer/.composer/vendor/bin/phpmetrics"
	reportPath string = "/tmp/phpmetrics.xml"
	reportFormat string = "REPORT"
)

func main() {
	// Get the CO_DATA from environment parameter "CO_DATA"
	data := os.Getenv("CO_DATA")
	keys := []string{
		"git-url",
		"path",
		"exclude",
	}
	codata := map[string]string{}

	if err := input.HandleInput(data, keys, codata); err != nil {
		fmt.Fprintf(os.Stderr, "[COUT] Handle input error: %s\n", err.Error())
		fmt.Fprintf(os.Stdout, "[COUT] CO_RESULT = %s\n", "false")
		os.Exit(1)
	}

	if err := git.Clone(codata["git-url"], basePath); err != nil {
		fmt.Fprintf(os.Stderr, "[COUT] Clone the repository error: %s\n", err.Error())
		fmt.Fprintf(os.Stdout, "[COUT] CO_RESULT = %s\n", "false")
		os.Exit(1)
	}

	if codata["path"] == "" {
		codata["path"] = "."
	}
	exclude := fmt.Sprintf("--exclude=%v", codata["exclude"])

	cmd := exec.Command(baseCommand, codata["path"], exclude,"--report-violations=/tmp/report.xml")
	cmd.Dir = basePath
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr

	if err := cmd.Run(); err != nil {
		fmt.Fprintf(os.Stderr, "[COUT] Create report error: %s\n", err.Error())
		fmt.Fprintf(os.Stdout, "[COUT] CO_RESULT = %s\n", "false")
		os.Exit(1)
	}

	file.StdoutAll(reportPath, reportFormat)

	fmt.Fprintf(os.Stdout, "[COUT] CO_RESULT = %s\n", "true")
	os.Exit(0)
}

File: component/php/Analysis/phpmetrics/src/component/Phpmetrics.php
<?php
class Phpmetrics {
    const reportPath = "/tmp/phpmetrics.xml";
    const reportFormat = "REPORT";
    
    public static function exec($input) {
        try {
            Git::clone($input[GIT_REPO]);
        } catch (Exception $ex) {
            stderrln("Git clone error: " . $ex->getMessage());
            stderrln("[COUT] CO_RESULT = false");
        }

        try {
            $cmd = "/root/.composer/vendor/bin/phpmetrics ";

            if ($input["path"] == "" || $input["path"] == null) {
                $input["path"] = ".";
            }

            $cmd = "$cmd " . $input["path"];

            $params = [
                "exclude"
            ];

            foreach ($params as $value) {
                if ($input[$value] != "") {
                    $cmd = "$cmd --$value=" . $input[$value];
                }
            }

            $params_bool = [
                "ignore-annotations",
            ];

            foreach ($params as $value) {
                if ($input[$value] == "true") {
                    $cmd = "$cmd --$value";
                }
            }

            $cmd = "$cmd --report-violations=" . self::reportPath;

            $lastLine = system("cd " . WORK_DIR . " && $cmd", $result);
            
            if ($result != 0) {
                throw new Exception();
            }

            stdoutReport(self::reportPath, self::reportFormat);
            stdoutln("[COUT] CO_RESULT = true");
        } catch (Exception $ex) {
            stderrln("[COUT] CO_RESULT = false");
        }
    }
}
?>